keps/sig-testing/1746-stage-e2e-test-framework/README.md

```diff
@@ -0,0 +1,610 @@
+  is used in inadequate e2e test framework sub-packages.
+* The e2e test framework will be refactored as needed strictly to remove its
+  dependence from other internal Kubernetes packages. It is not the goal of
+  this enhancement to propose an entire redesign of the e2e test framework.
+
+<!--
+What is out of scope for this KEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+In order to cleanup and refactor the e2e test framework we will have to
+decouple the e2e test framework from internal Kubernetes packages and to
+organize test code in e2e tests and in the e2e test framework.
+
+### User Stories
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+I, as a core Kubernetes contributor, want a e2e test framework that is well
+organized and containsa core of functions that will help me write clean and
+maintainable e2e tests.
+
+#### Story 2
+
+I, as a maintainer of a project running on top of Kubernetes (e.g. external cloud providers, CSI drivers), want to use the
+same infrastructure and tools used by core Kubernetes without having to import
+core Kubernetes.
+
+### Managing Internal Kubernetes Packages
+
+To decouple the e2e test framework from internal packages we essentially have
+to prevent any from `k8s.io/kubernetes/pkg` and from any other package in
+`k8s.io/kubernetes/test` that aren't under
+`k8s.io/kubernetes/test/e2e/framework`.
+
+These core dependencies are introduced in the following ways
+1. as primitives for test utility functions to manage Kubernetes resources and objects
+2. as data structures to manage and consume data generated by a Kubernetes component
+3. as variables and constants
+
+To resolve the dependencies introduced by bullet point 1, we plan to move or
+rewrite e2e test framework functions using official packages and APIs to match
+the current desired behaviour.
+For the most part this should be straightforward and it will only cause issues
+when one of the other two following conditions are present.
+
+Bullet point number 2 is present in cases where e2e tests involve collecting
+metrics from unpublished APIs.
+As an example, see
+[kubeletstatsv1alpha1](https://github.com/kubernetes/kubernetes/blob/ba3bf32300574d86c98657981c96ca609de787a2/test/e2e/framework/resource_usage_gatherer.go#L37)
+being imported to collect Kubelet statistics.
+We want to prevent this kind of internal dependencies in order to ensure that
```

Comment:
This seems like it runs counter to ensuring that kubernetes itself is well tested and is probably my biggest complaint with this effort.

Presumably these unpublished APIs exist and are unpublished for a reason, such as being useful for testing kubernetes itself, but not being designed for external consumption.

Reply:
I think there is an argument for saying that e2e tests should test workflows that users can follow but i think that line of discussion would be bikeshedding.

To the practical point of testing Kubernetes internals, we would still be able to do so.
The e2e test framework itself is supposed to hold "configuration information" (i.e., where is my cluster for e2e tests, where is etcd). e2e framework subpackages are meant to hold useful utility functions used by multiple e2e test pacakges.

If we dont allow unpublished APIs within the e2e framework, code that uses them can still be used by test packages.
If someone wanted to do an e2e test that relies on an internal Kubelet API they would be able to do so by adding utility functions somewhere like `k/k/test/e2e/kubelet` (just an example).